src/pykeen/nn/emb.py

```diff
@@ -580,3 +581,246 @@ def forward(
+        composition: Hint[CompositionFunction] = None,
+        activation: Hint[nn.Module] = nn.Identity,
+        activation_kwargs: Optional[Mapping[str, Any]] = None,
+    ):
+        super().__init__()
+
+        # entity-relation composition
+        self.composition = composition_resolver.make(composition)
+
+        # message passing weights
+        self.w_loop = nn.Parameter(data=torch.empty(input_dim, output_dim))
+        self.w_fwd = nn.Parameter(data=torch.empty(input_dim, output_dim))
+        self.w_bwd = nn.Parameter(data=torch.empty(input_dim, output_dim))
+
+        # linear relation transformation
+        self.w_rel = nn.Linear(in_features=input_dim, out_features=output_dim, bias=use_relation_bias)
+
+        # layer-specific self-loop relation representation
+        self.self_loop = nn.Parameter(data=torch.empty(1, input_dim))
+
+        # other components
+        self.drop = nn.Dropout(dropout)
+        self.bn = nn.BatchNorm1d(output_dim)
+        self.bias = Bias(output_dim) if use_bias else None
+        self.activation = activation_resolver.make(query=activation, pos_kwargs=activation_kwargs)
+
+        # initialize
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        """Reset the model's parameters."""
+        for w in (
+            self.w_loop,
+            self.w_fwd,
+            self.w_bwd,
+            self.w_rel,
+            self.self_loop,
+        ):
+            nn.init.xavier_uniform_(w)
+        self.bias.reset_parameters()
+
+    def message(
+        self,
+        x_e: torch.FloatTensor,
+        x_r: torch.FloatTensor,
+        triples: Optional[Tuple[torch.LongTensor, torch.LongTensor, torch.LongTensor]],
+    ) -> torch.FloatTensor:
+        if triples is None:
+            # compose
+            m = self.composition(x_e, self.self_loop)
+            # transform (no dropout)
+            return m @ self.self_loop
+
+        # compose
+        h, r, t = triples
+        m = self.composition(x_e[h], x_r[r])
+        # transform
+        m = m @ self.w
+        # TODO: normalization
+        # self.in_norm = self.compute_norm(self.in_index, num_ent)
+        # aggregate by sum
+        x_e = torch.zeros_like(x_e).index_add(dim=0, index=t, source=m)
+        return self.drop(x_e)
+
+    def forward(
+        self,
+        x_e: torch.FloatTensor,
+        x_r: torch.FloatTensor,
+        triples: torch.LongTensor,
+    ) -> Tuple[torch.FloatTensor, torch.FloatTensor]:
+        # split
+        h, r, t = triples.t()
+
+        # entity transformation
+        x_e = sum((
+            self.message(
+                x_e=x_e,
+                x_r=x_r,
+                triples=triples_,
+            ),
+            for triples_ in (
+            (h, 2 * r, t),  # forward; todo make sure this is the correct interpretation of inverse relations
+            (t, 2 * r + 1, h),  # backward
+            None,  # self-loop
+        )
+        )) / 3
```

Comment:
3??? where does this number come from

Reply:
To compute the mean over the 3-element list inside

Reply:
thanks for clarification. Would be nice to add this as a comment in the code, since this is a hella complicated one liner